espnet2/train/gan_trainer.py

```diff
@@ -0,0 +1,470 @@
+                                continue
+                            scaler.step(optimizer)
+                            scaler.update()
+
+                for iopt, optimizer in enumerate(optimizers):
+                    # NOTE(kan-bayashi): In the case of GAN, we need to clear
+                    #   the gradient of both optimizers after every update.
+                    optimizer.zero_grad()
+
+                # Register lr and train/load time[sec/step],
+                # where step refers to accum_grad * mini-batch
+                reporter.register(
+                    {
+                        f"optim{optim_idx}_lr{i}": pg["lr"]
+                        for i, pg in enumerate(optimizers[optim_idx].param_groups)
+                        if "lr" in pg
+                    },
+                )
+                reporter.register(
+                    {f"{turn}_train_time": time.perf_counter() - turn_start_time}
+                )
+                turn_start_time = time.perf_counter()
+
+            reporter.register({"train_time": time.perf_counter() - start_time})
+            start_time = time.perf_counter()
+
+            # NOTE(kamo): Call log_message() after next()
+            reporter.next()
+            if iiter % log_interval == 0:
+                logging.info(reporter.log_message(-log_interval))
+                if summary_writer is not None:
+                    reporter.tensorboard_add_scalar(summary_writer, -log_interval)
+                if use_wandb:
+                    reporter.wandb_log()
+
+        else:
+            if distributed:
+                iterator_stop.fill_(1)
+                torch.distributed.all_reduce(iterator_stop, ReduceOp.SUM)
+
+        return all_steps_are_invalid
+
+    @classmethod
+    @torch.no_grad()
+    def validate_one_epoch(
+        cls,
+        model: torch.nn.Module,
+        iterator: Iterable[Dict[str, torch.Tensor]],
+        reporter: SubReporter,
+        options: TrainerOptions,
+        distributed_option: DistributedOption,
+    ) -> None:
+        assert check_argument_types()
+        ngpu = options.ngpu
+        no_forward_run = options.no_forward_run
+        distributed = distributed_option.distributed
+
+        model.eval()
+
+        # [For distributed] Because iteration counts are not always equals between
+        # processes, send stop-flag to the other processes if iterator is finished
+        iterator_stop = torch.tensor(0).to("cuda" if ngpu > 0 else "cpu")
+        for (_, batch) in iterator:
+            assert isinstance(batch, dict), type(batch)
+            if distributed:
+                torch.distributed.all_reduce(iterator_stop, ReduceOp.SUM)
+                if iterator_stop > 0:
+                    break
+
+            batch = to_device(batch, "cuda" if ngpu > 0 else "cpu")
+            if no_forward_run:
+                continue
+
+            for turn in ["generator", "discriminator"]:
+                forward_generator = turn == "generator"
```

Comment:
This line is the same as line 176. It can be written as a function.

Reply:
changed to one liner.